horovod/common/operations.cc

```diff
@@ -603,86 +661,136 @@ bool RunLoopOnce(HorovodGlobalState& state) {
     state.timeline.MarkCycleStart();
   }
 
-  auto response_list =
-      state.controller->ComputeResponseList(horovod_global.shut_down, state);
+  bool this_process_requested_shutdown = state.shut_down;
 
-  state.mark_cycles_in_timeline =
-      state.controller->MarkCyclesInTimelinePending();
-
-  // Get tensor name and size data for autotuning.
-  int64_t total_tensor_size = 0;
-  std::vector<std::string> tensor_names;
-  if (state.parameter_manager.IsAutoTuning()) {
-    total_tensor_size = horovod_global.tensor_queue.GetTensorDataForAutotuner(
-        response_list, tensor_names);
-  }
-
-  // Perform the collective operation. All nodes should end up performing
-  // the same operation.
-  int rank = state.controller->GetRank();
-  for (auto& response : response_list.responses()) {
-    if (!state.group_table.empty()) {
-      // Deregister any completed groups
-      state.group_table.DeregisterGroups(response.tensor_names());
+#if HAVE_MPI
+  if (state.dynamic_process_sets) {
+    // Initialize any newly added process set that has been registered by all
+    // Horovod processes.
+    if (global_mpi_context.IsEnabled()) {
+      state.process_set_table.InitializeRegisteredIfReady(global_mpi_context);
+      // Remove a process set that has been marked for removal by all Horovod
+      // processes.
+      state.process_set_table.RemoveMarkedProcessSetIfReady();
     }
+  }
+#endif // HAVE_MPI
 
-    LOG(TRACE, rank) << "Performing " << response.tensor_names_string();
-    LOG(TRACE, rank) << "Processing " << response.tensor_names().size()
-                     << " tensors";
-    PerformOperation(response, horovod_global);
-    LOG(TRACE, rank) << "Finished performing "
-                     << response.tensor_names_string();
+  std::unordered_map<int32_t, ResponseList> process_set_response_lists;
+  for (auto process_set_id : state.process_set_table.Ids()) {
```

Comment:
Would it be faster to interleave the `ComputeResponseList` calls and `PerformOperation` calls here, rather than doing them in two separate loops? This would allow more rapid scheduling of collective ops and potentially less GPU idle time, since you aren't waiting for coordination of every process group in a big sequential chunk before launching collectives. 

Caveat: There is the WAIT_FOR_DATA spin loop that will hold back the background thread on every `PerformOperation` call, however, I am planning on posting a PR soon to modify that to use a more asynchronous `cudaStreamWaitEvent` to manage the data dependency. Seems like this feature could benefit from that.

Reply:
Yes, I think you are right that interleaving those two loops has the potential to reduce latency. Also I believe there is nothing left in `ComputeResponseList` (except for process set 0) that would make use of the global `Controller`, so there is reason to believe that this should not cause any dead locks. I've made the change in my last commit for today.

> Caveat: There is the WAIT_FOR_DATA spin loop that will hold back the background thread on every PerformOperation call

I see. Would that matter in typical situations where there is one global process set (e.g., `[0,1,2,3]`) and the other process sets are non-overlapping (e.g., `[0,1]` and `[2,3]`)? If the background thread is blocked in processes 0 and 1 because of a `PerformOperation` call for process set `[0,1]`, the processes in `[2,3]` should not be affected and can call `ComputeResponseList` right away.

> however, I am planning on posting a PR soon to modify that to use a more asynchronous cudaStreamWaitEvent to manage the data dependency. Seems like this feature could benefit from that.

That sounds nice!

Reply:
> If the background thread is blocked in processes 0 and 1 because of a PerformOperation call for process set [0,1], the processes in [2,3] should not be affected and can call ComputeResponseList right away.

Yes, that is right. The background thread for `[2,3]` should not be impacted by a block on `[0,1]` if those are the only sets in use (non-overlapping). I was thinking more about the cases where there is overlap in the process sets, which I think will also be a common use case, since at some point you probably need to aggregate results/do some computation across your disjoint process sets using the global process set or another overlapping process set. 

Thanks for making the change to interleave the operations.